src/watchdog.js

```diff
@@ -0,0 +1,401 @@
+/**
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module watchdog/watchdog
+ */
+
+/* globals console, window */
+
+import mix from '@ckeditor/ckeditor5-utils/src/mix';
+import EmitterMixin from '@ckeditor/ckeditor5-utils/src/emittermixin';
+import { throttle } from 'lodash-es';
+import CKEditorError from '@ckeditor/ckeditor5-utils/src/ckeditorerror';
+import areConnectedThroughProperties from '@ckeditor/ckeditor5-utils/src/areconnectedthroughproperties';
+
+/**
+ * A watchdog for CKEditor 5 editors.
+ *
+ * It keeps an {@link module:core/editor/editor~Editor editor} instance running.
+ * If a {@link module:utils/ckeditorerror~CKEditorError `CKEditorError` error}
+ * is thrown by the editor, it tries to restart the editor to the state before the crash. All other errors
+ * are transparent to the watchdog. By looking at the error context, the Watchdog restarts only the editor which crashed.
+ *
+ * Note that the watchdog does not handle errors during editor initialization (`Editor.create()`)
+ * and editor destruction (`editor.destroy()`). Errors at these stages mean that there is a serious
+ * problem in the code integrating the editor and such problem cannot be easily fixed restarting the editor.
+ *
+ * Basic usage:
+ *
+ * 		const watchdog = Watchdog.for( ClassicEditor );
+ *
+ * 		watchdog.create( elementOrData, editorConfig ).then( editor => {} );
+ *
+ * Full usage:
+ *
+ *		const watchdog = new Watchdog();
+ *
+ *		watchdog.setCreator( ( elementOrData, editorConfig ) => ClassicEditor.create( elementOrData, editorConfig ) );
+ *		watchdog.setDestructor( editor => editor.destroy() );
+ *
+ *		watchdog.create( elementOrData, editorConfig ).then( editor => {} );
+ *
+ * Other important APIs:
+ *
+ *		watchdog.on( 'crash', () => { console.log( 'Editor crashed.' ) } );
```

Comment:
There is no such event fired. 'error' is fired instead, but as mentioned I am not sure if it is needed.